src/test/java/com/es/phoneshop/web/ProductDetailsPageServletTest.java

```diff
@@ -0,0 +1,71 @@
+package com.es.phoneshop.web;
+
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+import com.es.phoneshop.service.ProductServiceImpl;
+import jakarta.servlet.RequestDispatcher;
+import jakarta.servlet.ServletConfig;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.Mock;
+import com.es.phoneshop.exceptions.ProductNotFoundException;
+import com.es.phoneshop.model.product.Product;
+import org.mockito.MockitoAnnotations;
+
+import java.util.Optional;
+
+public class ProductDetailsPageServletTest {
+
+    private ProductDetailsPageServlet servlet;
+
+    @Mock
+    private ProductServiceImpl productService;
+
+    @Mock
+    private HttpServletRequest request;
+
+    @Mock
+    private HttpServletResponse response;
+
+    @Mock
+    private RequestDispatcher requestDispatcher;
+
+    @Mock
+    private ServletConfig servletConfig;
+
+    @Before
+    public void setup() throws Exception {
+        MockitoAnnotations.initMocks(this);
+        servlet = new ProductDetailsPageServlet();
+        servlet.init(servletConfig);
+        when(request.getRequestDispatcher(anyString())).thenReturn(requestDispatcher);
+    }
+
+    @Test
+    public void givenValidProductId_whenDoGet_thenValidAttributeAndForward() throws Exception {
+        ProductServiceImpl productService1 = ProductServiceImpl.getInstance();
+        Product product = new Product();
+        String productId = String.valueOf(product.getId());
+        productService1.save(product);
```

Comment:
you don't need to rely on real implementation of save() while testing servlet.doGet()
In your servlet.doGet() there is only one call of productService: productService.getProduct()
While testing servlet you need to verify that the request stores correct product and that requestDispatcher forwarded the request